sound/soc/sof/pcm.c

```diff
@@ -531,9 +536,16 @@ static int sof_pcm_close(struct snd_pcm_substream *substream)
 
 	/* update active pcm counts */
 	sdev->pcm_active--;
-	if (spcm->stream[substream->stream].d0i3_compatible)
+	if (spcm->stream[substream->stream].d0i3_compatible) {
 		sdev->lp_pcm_active--;
 
+		/* may be possible to exit from d0i3 if, not forcing */
+		snd_sof_set_dsp_d0i0(sdev, false, true);
+	} else {
+		/* may be possible to enter d0i3 */
+		snd_sof_set_dsp_d0i3(sdev);
+	}
+
```

Comment:
I actually would rather comment to the commit message of this patch - it looks very much like re-inventing runtime-PM to me, can we convert this to use the standard RT-PM APIs?

Reply:
yes, it feels like re-invention but no it's not. pm_runtime is only about DO-D3. When the device is in D0i3 (which is platform-specific), pm_runtime considers the device as 'active'. So indeed we need to implement a custom usage/refcount and decision making. It's not necessarily elegant but it's not generic either so the ability to rely on a common framework for stuff that's hardware/firmware specific is limited.

Reply:
Hi @lyakh and @plbossart I think using RT-PM API is doable, it only define runtime_status as either suspended or active, not limit them to be D3/D0, we can make it suspended==d0i3 and active==d0i0.

I did a prototype code to have a quick try today, it works fine. What I did is:
Introduce a new platform device(child of pci device) with PM ops(runtime suspend/resume),
runtime get at pcm_open() and put at pcm_close() for normal substream only.

@plbossart if you can accept this solution, I will clean up and update it to the PR tomorrow.

Reply:
@keyonjie I see @plbossart's point that RTPM is only about D0 / D3, but still doing what we want to do "from scratch" doesn't seem very pretty to me. At the very least I'd look at using kref instead of doing manual use-counting. But even better I'd talk to Linux PM gurus. Hint:: @rjwysocki is at Intel